### ISSUE TITLE
QUESTION: Designing skins and using the  Stronghold Image Toolbox

### ISSUE THREAD

**Issue by Wolfram92** [Sh0wdown/UnofficialCrusaderPatch#206]
This is more of a general question which is not dircetly related to the patch.

I've looked into the Stronghold Image Toolbox which lets you extract the skins and animation files for SC1 buildings and NPCs. As far as I unterstand, skins for building are literally just .bmp's, while animations work like a .GIF. Replacing the skins for buildings seems fairly easy, animation seems quite buggy.

Now my question: I have no idea how the Toolbox really works, and if it's possible to edit the Images or just to extract and replace them.

I could imagine doing some skin work for a new building like a mosque for example, but I'm not sure which format I would have to use (bmp I guess?), which size, background colour etc.

I would make a general sketch and then I could paint it in a real painting program (not paint).

![moschee](https://user-images.githubusercontent.com/46036962/50740951-f0238f00-11f6-11e9-9f4b-ef8e5e11eb4e.jpg)

So this would be an example of a mosque, which I have done in paint so I know it looks rather crappy :P 

I can imagine doing a far better version in Krita (paint program). So the 'painting' (creating a skin) is not the problem here, just implementing.

If someone could help me out with that part, then I could do a few skins:)

**Comment by Monsterfisch** [comment]
Image Toolbox only lets you extract and replace images from/of the game files in order to do it you have to use the edit button in the top after selecting the tile(some tiles can also be selected together and exported as a group and imported as well) 

I did a few changes myself with the program to update the stronghold Europe mod to add some fancy [roofs to the towers](https://i.imgur.com/Rboplap.png) but encountered issues while trying to fix some pink pixels in the moat and the darkness of the cracks on broken buildings where the game started displaying horizontal artifacts for some reason =/ (don't have a screenshot of it but essentially it was horizontal lines  with random color noise inside them with the tiles themselves and tiles other tiles like small waterfalls that are located in the same file as the moat tiles.

**Comment by Wolfram92** [comment]
Ok, so in this example (picture below) I open the animation file for the fletcher which is composed of still frames which together for the complete animation. I selected the first image and now I could replace with an image of my own, right? But what kind of file do I need and how should it look like?

![toolbox1](https://user-images.githubusercontent.com/46036962/50741898-2adff400-1204-11e9-8a05-b9592d4ce81d.jpg)

In the second case I opened the image of the church, which is just a skin, without an animation. So I guess I just have to replace the image here, right? Can you elaborate how you designed the roofs?

![toolbox2](https://user-images.githubusercontent.com/46036962/50741919-5d89ec80-1204-11e9-801f-7be5ca533014.jpg)

**Comment by Monsterfisch** [comment]
The file format you need to import as far as I know is always .bmp .

for the roofs i made use of the roof feature for those two towers in crusader essentially changeing these files
![image](https://user-images.githubusercontent.com/24592166/50742046-65e32700-1206-11e9-93d7-f780d6210bb4.png)

to these 

![image](https://user-images.githubusercontent.com/24592166/50742056-827f5f00-1206-11e9-87b2-d00420931553.png)

the roof texture in this case is a smlightly modified version of a workshop and the healer building.


as for how the file should look loike you will see after you export the image which can be done by clicking the save button in the edit window. you than edit the file to your liking and than import it with the inject button. as i mentioned before there are some issues with graphic glitches that i dont know how to fix so if you find a way to bypass those I will have very open ears ;)

**Comment by Wolfram92** [comment]
Aaaah, now I got, thank you very much :) I thought 'saving' meant replacing the image, but this means that one has to save the image so that it will become availabe for editing, right?

I will play around a bit (I think I start with buildings and the aforementioned mosque) and if I happen to find a solution to those graphic glitches I will share it here;)

**Comment by Monsterfisch** [comment]
this is the kind of glitch i am talking about btw 
![image](https://i.imgur.com/NUZNDAn.gif)

![image](https://i.imgur.com/lBW2f2U.png)

From what i can tell at least the game is trying to load a bugged out frame at the end of the file that contains the fixed moat textures i tried a couple of different bmp settings but to no success yet might even be a problem with the image replacer tool itself?

**Comment by Wolfram92** [comment]
> From what i can tell at least the game is trying to load a bugged out frame at the end of the file that contains the fixed moat textures i tried a couple of different bmp settings but to no success yet might even be a problem with the image replacer tool itself?

According to the developer of Toolbox this is a known error which occurs for animations. He hasn't figured it out and unfortunately he stopped working on Toolbox alltogether.

http://stronghold.heavengames.com/downloads/showfile.php?fileid=4474

This is why I don't wanna touch animations right now, since it requires editing many images and it might look like crap in the end.

**Comment by Monsterfisch** [comment]
damn... =/ maybe showd0wn  will eventually find the time to look into that as well not that it would be a high priority through ofc who knows at this point but at least static things work out so can at least edit those

**Comment by Sh0wdown** [comment]
Have you tried simply adding more frames to the animation?
![image](https://user-images.githubusercontent.com/24592166/50742752-85328200-120f-11e9-92ce-c5d13ae58631.png)

**Comment by Monsterfisch** [comment]
I couldn't find a way to add more frames with the program and my knowledge in editing the files directly is also insufficient =/

Thing is, this is exactly the amount of frames the game uses normally (I am not removing existing frames)
its just really seems like after injecting a file it just skips a file for some reason. Almost like it commented a section of the code by accident.

**Comment by Wolfram92** [comment]
Hm, so today I inserted an incomplete prototype of a big mosque, analogous to a cathedral, into my SCH folder and placed one in the map editor just to get a sense of perspective.

And then this glitch happened:

![screen_capture_001](https://user-images.githubusercontent.com/46036962/50800182-6d2a3380-12df-11e9-97a7-8677ca860f79.jpg)

Have you encoutered this before? The background-color of my .bmp is the exact same of the original, but this raster like dots are being displayed here.

**Comment by Heroesflorian** [comment]
I have never tried replacing ingame textures with that tool so far, but I have noticed when changing the player "avatar" crests file (same magenta background color), it won't work properly depending on what encoding scheme you choose when saving the .bmp file.

**Comment by Wolfram92** [comment]
Well, that might explains it. I saved it with paint this (was faster) and maybe the encoding was wrong. I just try it again and see what happens.

**Comment by Wolfram92** [comment]
Almost fixed it.

I think just the process of extracting files and then resaving them, changes something so that the graphical glitch above occurs But changing the backgroundcolor to the same color in the Toolbox will make the background transparent.

![screen_capture_001](https://user-images.githubusercontent.com/46036962/50803919-15de9000-12ec-11e9-9c2e-d67d3ac1dd47.jpg)


In this case I used red but it is still glitchy. I managed to insert just a slightly modified cathedral into the .gm file without any problems by changing the backgroundcolor.

**Comment by Lolasik011** [comment]
WOW !

**Comment by Wolfram92** [comment]
Thanks:)

T will try to make skins for a big mosque (cathedral), normal mosque (church) and the chapelle, I would replace with a minaret, since there is no other equivalent as far as I know and it would fit size wise. I looked at some pictures of mosque and will use them as an inspiration.

I think it will take me a couple of weeks for one skin (never did that beforce actually) since the biggest challenges are:

-perspective drawing
-drawing in general because the resolution is so damn small that it's hard to draw details
-and drawing in a way that it still fits SC1 style.

When I am finished with all three I will open a new issue for them to get some feedback and if I am succesfull I will also look into animations perhaps, since a mosque for example would also need something like an Iman instead of a priest.

**Comment by Monsterfisch** [comment]
can always paint in higher resolution and than downscale it to fit later if you want to have an easier time detailing (always a good idea to reuse elements that already exist to save time and helps keeping it in line with the aesthetics btw ;) )

**Comment by Wolfram92** [comment]
> can always paint in higher resolution and than downscale it to fit later if you want to have an easier time detailing (always a good idea to reuse elements that already exist to save time and helps keeping it in line with the aesthetics btw ;) )

I'm just afraid that painting in a higher resolution and then downsizing would probably be bad for the details?

Yes, I definitely try to reuse us much as possible but some things I have to paint myself, like the little roof which turned out to be quite good I think:P My goal is to get a mosque which is somewhat decent looking and then I'll try to give it some finetuning by getting it as close to the SC1 style as possible.

**Comment by Monsterfisch** [comment]
when downsizing detail sure will get lost but at the same time, it also compresses detail in a way your brain probably wouldn't so it can go either way combining different resizing methods I am sure you will find the perfect way to get what you want.

**Comment by PitchNeeded** [comment]
Really looking forward to seeing this mosque in the game (and being able to place it when playing as an Arabian lord) once it is finished . :D

**Comment by PitchNeeded** [comment]
After the mosques, the next thing should be changing the ale production for Arabian lords into something else. 

Hops farms could be changed into spice farms (the spices being used to flavour food), the breweries changed into cooking kitchens and the inns changed into restaurants which serve better food than the usual granary supplies.

**Comment by Wolfram92** [comment]
>Hops farms could be changed into spice farms (the spices being used to flavour food), the breweries changed into cooking kitchens and the inns changed into restaurants which serve better food than the usual granary supplies.
 
Do you have a certain spice in mind? It has to be something similar in height to hops, because otherwise I would have to Change the complete Animation which I can't do.

Another Problem would the breweries, which would need to be repurposed too.

I will look into it after doing the mosques and statues, but I think it will be quite complicated.

**Comment by Sh0wdown** [comment]
> > Hops farms could be changed into spice farms (the spices being used to flavour food), the breweries changed into cooking kitchens and the inns changed into restaurants which serve better food than the usual granary supplies.
> 
> Do you have a certain spice in mind? It has to be something similar in height to hops, because otherwise I would have to Change the complete Animation which I can't do.
> 
> Another Problem would the breweries, which would need to be repurposed too.
> 
> I will look into it after doing the mosques and statues, but I think it will be quite complicated. Editing animations with Toolbox can be bugged, so it might not be possible at all.

You could do something like a Shisha Bar

**Comment by Monsterfisch** [comment]
If only animation files did not cause these visual glitches =/ 

would it be possible to make dry moats btw? 
Idea is that when you paint a moat on higher terrain (which currently isnt possible) your units would instead of wet one dig up a dry moat that behaves the same way as normal moat just beeing dry? 

I can easily get the terrain tiles ready for that.
Here is a little test tile I did in 10 min 
![image](https://user-images.githubusercontent.com/24592166/50973720-da111980-14e9-11e9-8888-6eb1000a045f.png)

**Comment by Monsterfisch** [comment]
I would argue it's not that important to differentiate between  ale or fermented juice at least as long as we don't know a better way to replace/import new graphics files without them causing issues

**Comment by Wolfram92** [comment]
> I would argue it's not that important to differentiate between ale or fermented juice at least as long as we don't know a better way to replace/import new graphics files without them causing issues

Changing the hop farm to grapes would be quite easy I think, the issue might be with the farmer if I replace his little basket full of hops with a basket full of wine.

I will try the Imam first because that should be the easiest thing to change and then I'll see if any glitches will occur.

**Comment by PitchNeeded** [comment]
> Changing the hop farm to grapes would be quite easy I think, the issue might be with the farmer if I replace his little basket full of hops with a basket full of wine.

And of course, get rid of the drunks too if the player using the inn is an Arabian lord.

**Comment by Wolfram92** [comment]
> And of course, get rid of the drunks too if the player using the inn is an Arabian lord.

For the Arabs they should just be removed completely, since their animation can't be used for anything else :P

**Comment by Heroesflorian** [comment]
Well, if Arabian lords can have +8 popularity bonus without annyoing drunkards rolling around, then that is a decisive gameplay advantage for human players! :D 

> I'm not sure if a shisha bar fits timewise :D I read though that Arabs used to drink 'fermented fruit juice', so maybe I could replace hops with grapes

Hmm... one could consider opium, possibly... although poppy fields would probably fit better with wheat farms than hops farms... meh.

**Comment by PitchNeeded** [comment]
That dry moat idea sounds good Monsterfisch! It should be quite deep and steep looking to fit in with the fact it is not passable by units without filling it in!

Although if you want to be realistic, I think maybe assassins should be able to cross a dry moat. They would however have to deploy their rope to go down it and back up at the other side which would slow them down a lot and give the enemy archers more time to fire at them.

**Comment by Monsterfisch** [comment]
making it look deep shouldn't be a problem not sure how much the game mechanic could be altered through I personally would just make it work like the normal moat you know so assassins are not so overpowered when the castle isn't located on the low level where it can have moat.

**Comment by PitchNeeded** [comment]
> I will try the Imam first because that should be the easiest thing to change and then I'll see if any glitches will occur.

If you and others on here can make the imam work with animations then that maybe raises the possibility of introducing new military units in the future which would be brilliant, which would really freshen up the game and also can be used for certain AIs to make them more unique.

Here's some ideas:

- Camel raiders (Arabian unit): These would look brilliant, they wouldn't be as fast as horse archers of course but a bit better armoured and could be a decent raiding and counterattacking unit. They would be a little bit cheaper to hire than a horse archer. In terms of AIs, Saladin and Wazir could support their horse archer raiders with some of these. 

The existing Stronghold Crusader camel model could be modified for the creation of this new unit.

- Desert axeman: Again these would look amazing, the woodcutter could be partially retextured to create this unit. They would be faster than an Arabian swordsmen but not as well armoured, but cheaper to hire. A new decently strong Arabian melee unit would be great, at the minute there is only Arabian swordsmen and assassins.

A Crusader axemen would also be great but this would be harder to create and would require a new option at the blacksmith to create axes.

- A crusader light cavalry option would be good. Spears, leather armour and a stables could allow the creation of these, they would be decent raiders and counter attackers, using their spears as melee weapons, or (I know it would very hard to implement though) being able to throw them at enemies. Similar armour and speed to the horse archer. They would be cheaper than knights, the stable horse limit would ensure though that not too many could be created.

- More castle defences: rolling logs or rock baskets on walls would be excellent. Rock baskets would be easier to create I think, the mangonel stone animations could be used for these when they are triggered.

**Comment by Wolfram92** [comment]
>If you and others on here can make the imam work with animations then that maybe raises the possibility of introducing new military units in the future which would be brilliant, which would really freshen up the game and also can be used for certain AIs to make them more unique. :D

Well, that might be possible for some time in the future but at the moment I'll try working on existing buildings and animations first, doing a whole new unit like a camel raider could be possible if one takes a slightly modified horse archers and puts him on the camel, but that unit would be super slow because the camels are always just walking.

>More castle defences: rolling logs or rock baskets on walls would be excellent. Rock baskets would be easier to create I think, the mangonel stone animations could be used for these when they are triggered.

That is probably easier because there is already a similar animation for the pitch throwing engineers and their pot could be replaced with a basket of rocks.

For the other suggestions there is the problem that there are no existing animations for them at all (like rolling logs), so this would be not just editing an existing animation but require doing a completely new one, which is very difficult.

I guess my big mosque will be finished the coming week and then I will try a test run.

One challenge is also to make it somehow possible that mosques and churches can be built at the same since at the moment it's only possible to replace the churches with mosques. I can't do that because I have no idea about modding :P

**Comment by PitchNeeded** [comment]
> Well, that might be possible for some time in the future but at the moment I'll try working on existing buildings and animations first, doing a whole new unit like a camel raider could be possible if one takes a slightly modified horse archer and puts him on the camel, but that unit would be super slow because the camels are always just walking.

Would it be possible to speed up the walking animations for the camel for the camel raider so they are running? :D

**Comment by Monsterfisch** [comment]
I would recommend not getting your hopes up right now considering we only have one person here who is actually able to do the coding and adding new units to the game requires alot of work appart from the visuals

**Comment by PitchNeeded** [comment]
What about having bridges across rivers on maps instead of just fords?

Bridges would look great, and also it would give the possibility of having battles developing at bridges in a game, which would be good fun.

**Comment by Monsterfisch** [comment]
I too like the idea of bridges or at least arches of sorts (detailing related and gameplay) in the past I have tried to use gatehouses to simulate bridges but that is a mediocre solution at best thanks to horses being unable to walk over walls without tricking the ai the only issue I can see with this would be units crossing over one another with the pathfinding that could be an issue here. so as before hopes are great but possibly misplaced in terms of seeing it any time soon. consider how gatehouses work in the game i am sure based on how they worked in later titles of the game even firefly didn't figure out a way to do the pathfinding in the crusader engine to make it work =/

**Comment by Heroesflorian** [comment]
If you are talking about bridges as a "map feature" to be created in the map editor (and not during a running game), then I think that should be quite possible with (lowered) walls that could possibly be mixed with regualar ground to make it traversable for mounted units as well. 
Possibly even in combination with the good old "2000 rocks trick" that persists the looks of a tile (and makes them impassable) in order to get the railings of the bridge to be indestructible...

**Comment by Wolfram92** [comment]
So, I finished my big mosque as you can see in the pic below, but the problem is, that the Image Toolbox crops my image, although it has the exact same measurements as the original one. It even shows it right before injecting but then it gets messed up.

![moschee-entwurf4 3](https://user-images.githubusercontent.com/46036962/51502489-a6ca6680-1dd6-11e9-9110-1bd29a8ac489.jpg)

Here is a picture from Toolbox

![toolbox1](https://user-images.githubusercontent.com/46036962/51502572-e002d680-1dd6-11e9-9d79-d493663bbeb3.jpg)

And this is what it looks like ingame. I think the glitches (the red parts) can be fixed, but the main problem is the messed up image. Any idea what's the issue is here, @Monsterfisch ?

![mosque-ingame](https://user-images.githubusercontent.com/46036962/51502602-ff99ff00-1dd6-11e9-8b3d-ef7bc2fff6e3.jpg)

**Comment by Monsterfisch** [comment]
I mean it seems to work in-game apart from the red pixels or am I missing something?

**Comment by Monsterfisch** [comment]
to get rid of the pixels simply paint the area a bit more generous the overlap will not be noticeable later I had a similar issue in the past with the small tower and covering the pixels that were not in the original one fixed it

**Comment by Wolfram92** [comment]
Man, that didn't let me sleep, but I finally got it working.

So, for everybody who maybe also wants to edit some of the buildings, here's to important things to know:

- The backgroundcolor of your image, has to be the exact same background color from your color palette in the Image Toolbox

-At the outer lines of the building should be no color overlap at all, or otherwise there will be some ugly pixels in the picture.

-And, very important, the 'floor', the square where your building is built upon, has to be the exact same size as the original. This seems to be somehow hardwired into the gm. files. If the floor won't match the exact scaling of the original, then the lower part of the image will always glitch out and the background color will be visible.

I solved this by having the floor filling out the complete lower part of the image, unfortunately, now it's not a real square anymore. But for the moment I can live with that.

![mosque-ingame](https://user-images.githubusercontent.com/46036962/51506683-c9b44500-1ded-11e9-8c1d-23705ae47ab8.jpg)

**Comment by Grochuuu** [comment]
> Człowieku, to nie pozwoliło mi spać, ale w końcu udało mi się.
> 
> Tak więc, dla każdego, kto może chce edytować niektóre budynki, ważne jest, aby wiedzieć:
> 
> * Kolor tła obrazu musi być dokładnie tym samym kolorem tła z palety kolorów w Przyborniku obrazu
> * Na zewnętrznych liniach budynku nie powinno w ogóle nakładać się kolorów, w przeciwnym razie na obrazie pojawią się brzydkie piksele.
> * I, co bardzo ważne, "podłoga", kwadrat, na którym zbudowany jest twój budynek, musi być dokładnie takiej samej wielkości jak oryginał. Wydaje się, że jest to w jakiś sposób podłączone do gm. pliki. Jeśli posadzka nie będzie pasować do dokładnego skalowania oryginału, wówczas dolna część obrazu będzie zawsze wyskakiwać, a kolor tła będzie widoczny.
> 
> Rozwiązałem to poprzez wypełnienie podłogi kompletną dolną częścią obrazu, niestety, teraz to już nie jest prawdziwy kwadrat. Ale na razie mogę z tym żyć.
> 
> ![meczet-gra](https://user-images.githubusercontent.com/46036962/51506683-c9b44500-1ded-11e9-8c1d-23705ae47ab8.jpg)

I could ask for a download of this texture :) ? (sorry for my english )

**Comment by Grochuuu** [comment]
> Format pliku, który musisz zaimportować, o ile wiem, to zawsze .bmp.
> 
> dla dachów skorzystałem z funkcji dachu dla tych dwóch wież w krzyżowcu, zasadniczo zmieniając te pliki
> ![obraz](https://user-images.githubusercontent.com/24592166/50742046-65e32700-1206-11e9-93d7-f780d6210bb4.png)
> 
> do tych
> 
> ![obraz](https://user-images.githubusercontent.com/24592166/50742056-827f5f00-1206-11e9-87b2-d00420931553.png)
> 
> tekstura dachu w tym przypadku jest nieznacznie zmodyfikowaną wersją warsztatu i budynku uzdrawiającego.
> 
> jeśli chodzi o to, jak powinien wyglądać plik, zobaczymy po wyeksportowaniu obrazu, który można zrobić, klikając przycisk zapisu w oknie edycji. możesz edytować plik do swoich potrzeb i importować go za pomocą przycisku wstrzykiwania. jak wspomniałem wcześniej, są pewne problemy z graficznymi usterkami, których nie wiem jak naprawić, więc jeśli znajdziesz sposób na ominięcie tych, będę miał bardzo otwarte uszy;)

I could ask for a download of this textures ?
 (my english is bad)

**Comment by PitchNeeded** [comment]
> Man, that didn't let me sleep, but I finally got it working.
> 
> So, for everybody who maybe also wants to edit some of the buildings, here's to important things to know:
> 
> * The backgroundcolor of your image, has to be the exact same background color from your color palette in the Image Toolbox
> * At the outer lines of the building should be no color overlap at all, or otherwise there will be some ugly pixels in the picture.
> * And, very important, the 'floor', the square where your building is built upon, has to be the exact same size as the original. This seems to be somehow hardwired into the gm. files. If the floor won't match the exact scaling of the original, then the lower part of the image will always glitch out and the background color will be visible.
> 
> I solved this by having the floor filling out the complete lower part of the image, unfortunately, now it's not a real square anymore. But for the moment I can live with that.
> 
> ![mosque-ingame](https://user-images.githubusercontent.com/46036962/51506683-c9b44500-1ded-11e9-8c1d-23705ae47ab8.jpg)

That's lovely, you should place it in a complete castle in a sandbox game and upload a screenshot to let us see what it looks like in a castle! :D

**Comment by Grochuuu** [comment]
dajcie mi te textury do downloadingu pls

**Comment by Wolfram92** [comment]
I will place it in castle to give you an idea what it looks like. Later on I will also post a download link so that you can try it out for yourself:)

I will also add windows to the mosque and adapt the color a little bit to the other stronghold buildings, because at the moment it doesn't fit well in with the rest in my opinion.

**Comment by PitchNeeded** [comment]
An idea eventually:

New skins for all the AI lords, so for example the Rat lord looks like and is dressed like the Rat. Saladin and Richard would stay the same as they are.

**Comment by Gaaammmler** [comment]
@Wolfram92 i started to create my own Image Toolbox, it works properly well now and i can import the church and sea tiles correct, with the next update i realease bigger functionalty and Error handling :) 

![exported](https://user-images.githubusercontent.com/5760157/57475108-f2c0ee80-7293-11e9-9927-ea3dd9ae5d49.JPG)

**Comment by Wolfram92** [comment]
Wow very nice:) I have been very inactive in recent times because of work, but next week I'll have quite a lot of spare time.

I also had problems with the background glitching through in the Toolbox, so I yours can fix that hopefully:)

**Comment by Gaaammmler** [comment]
@Wolfram92 I created some Important infors in the Programm you can read(click on the info button).
You can donwload the Programm here: https://github.com/Gaaammmler/Gm1KonverterCrossPlatform/releases

**Comment by ISTRONGHOLD** [comment]
@Wolfram92 please give a download link for your mosque

**Comment by ISTRONGHOLD** [comment]
much appriciated

**Comment by mantaslukocius** [comment]
@PodeCaradox Hello do you still making something? im making high quality 3D models for game like Strong hold

**Comment by PodeCaradox** [comment]
Yea im still slowly working on this here: https://podecaradox.github.io/CastleSimWeb/
You need to use chrome for the demo

**Comment by mantaslukocius** [comment]
@PodeCaradox yes i tried it is amazing bro! how to contact you personally, hope you see some pictures i add. im making 3D stuff for fun, and not find direction what to do and love Stonghold Crusader, would be nice to help you to add animated NPC buildings trees. you are doing some amazing stuff. i can make 8 direction or more any sprite size animations with similar color and art style

**Comment by PodeCaradox** [comment]
Do you have Discord? If yes you can add me :)

**Comment by mantaslukocius** [comment]
added you
ziorke#3654 
some of animations, can make any size length . dont know about spinning, need to fix spring. i can make many troops from catapults and so on 
![fdghfgfggf](https://github.com/user-attachments/assets/cb628593-d762-4961-abe2-81e78b9ef026)
![gfhjgfhfghfghfghhf](https://github.com/user-attachments/assets/8683acd7-fdcf-4f14-aa22-c2d35029edd4)
![1720734323818 (1)](https://github.com/user-attachments/assets/c2e96d52-50ca-4f8f-a125-8c32d1178804)
